### ISSUE TITLE
`// format: off` doesn't preserve code from being rewritten

### ISSUE THREAD

**Issue by seanwentzel** [scalameta/scalafmt#1110]
- **Version**: 1.4.0

## Steps

Given code like this:
```scala
object Foo {
   // format: off
   val nbsp = "\u00A0"
   val test = s"${nbsp}\uD83D"
   // format: on
}
```

When I run scalafmt like this:

```bash
scalafmt --config <our_config> InputFile.scala
```

## Problem

Scalafmt formats code like this:
```scala
object Foo {
  // format: off
  val nbsp = "\u00A0"
  val test = s"${nbsp}D"
  // format: on
}
```
The diff would be expected without the format tags because of the bug https://github.com/scalameta/scalameta/issues/819

## Expectation

I would like the formatted output to be identical to the input because code with formatting turned off should be preserved exactly.

## Workaround

The only workaround is to structure our code to avoid the scalameta bug.

**Comment by olafurpg** [comment]
Thanks for reporting! This is a known scalameta bug with a fix in https://github.com/scalameta/scalameta/pull/1270

Unfortunately, we are blocked by #975 in order to enjoy that patch. Until we upgrade to scalameta v3.0 from v1.7 I'm afraid there is no workaround.

**Comment by olafurpg** [comment]
EDIT, the following workarounds might work

* replace unicode escape with `Int.toChar`
```scala
@ '\uD83D'.toInt
res1: Int = 55357

@ 55357.toChar
res2: Char = '\ud83d'

```

* exclude the source file containing the unicode escape characters with `project.excludeFilters`

**Comment by seanwentzel** [comment]
> exclude the source file containing the unicode escape characters with project.excludeFilters
This is a good workaround, thanks.

The issue I'm trying to raise here isn't so much the unicode bug, but the fact that there is a class of bugs for which `// format: off` tags don't protect code from getting rewritten. Excluding source files is a viable don't-need-to-understand-the-bug alternative, however. I'm happy to close this issue unless you want to keep it open to track this.

**Comment by olafurpg** [comment]
That is a good point. I might be possible to change the implementation of `// format: off` to slice the string contents of the positions instead of joining `token.syntax` as scalafmt currently does. I rarely use `// format: off` myself but anyone is welcome to take a look at contributing this if they are interested. The current implementation is all based around computing `Split` for each pair of non-whitespace tokens. For `// format: off`, we use the whitespace in the original source https://github.com/scalameta/scalafmt/blob/a9cc066127feca7384b03b516821e1840056d0db/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala#L219

A fix for this would require us to rethink how that is handled, for example by changing `FormatWriter` https://github.com/scalameta/scalafmt/blob/a9cc066127feca7384b03b516821e1840056d0db/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala 

I've renamed the issue title to the original title.

**Comment by olafurpg** [comment]
This has been fixed in #1170 by upgrading to scalameta v3.7. Opened #1175 to add test case to show I'm unable to reproduce on master.
